Jint surfing

BoZeng1997 opened the pull request:
A negative J2 invariant should never happen again. But I have not yet found out why the J integral over the boundary does not approach the value of the prescribed Gc.
In the folder /benchmark/AT2surfing, there are input deck for surfing boundary problem under AT-2 without irreversibility. As a standard phase-field fracture problem it still cannot return a close value to Gc.

hugary1995 commented on benchmarks/AT2surfing/submit.sh:
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
```
Remove this file from version control. You can use for example `git rm` for that.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+  Real K = _Lambda+2.0*_mu/3.0;
+  _F_surface[_qp] = _J2[_qp]/_mu + I1*I1/9.0/K - _ex_driving[_qp] - _temp;
+
+// if (std::isnan(_ex_driving[_qp]))
+// {
+//   std::cout << "_invar_1[_qp] = " << _invar_1[_qp] << std::endl;
+//   std::cout << "_invar_2[_qp] = " << _invar_2[_qp] << std::endl;
+//   std::cout << "_beta_0[_qp] = " << _beta_0[_qp] << std::endl;
+//   std::cout << "_beta_1[_qp] = " << _beta_1[_qp] << std::endl;
+//   std::cout << "_beta_2[_qp] = " << _beta_2[_qp] << std::endl;
+//   std::cout << "_beta_3[_qp] = " << _beta_3[_qp] << std::endl;
+//   std::cout << "_J2[_qp] = " << _J2[_qp] << std::endl;
+//   std::cout << "_ex_driving[_qp] = " << _ex_driving[_qp] << std::endl;
+// }
```
Remove these comments.

hugary1995 commented on tutorials/mode1_brittle_fracture/submit.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
Remove this file from version control.

hugary1995 commented on tutorials/mode1_cohesive_fracture/submit.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
Remove this file from version control.

hugary1995 reviewed (changes requested):
Please remove the scripts for submitting jobs. Also remove output files like `surf_AT2_Gc2L0.35del1.16_dt1e-2_J5b_14523984`.

hugary1995 commented:
Directly writing to standard streams (like std::cout) is prohibited. You can use them in your local branching for debugging purposes, but remember to remove those in a PR.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,244 @@
+    boundary_id_offset = 7
+    boundary_name_prefix = bottom_half
+  []
+  # [bottom_nonstitch]
+  #   type = BoundingBoxNodeSetGenerator
+  #   input = bottom_half
+  #   new_boundary = bottom_nonstitch
+  #   bottom_left = '0 0 0'
+  #   top_right = '5 0 0'
+  #   show_info = true
+  # []
```
Remove unused comments in the input files.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,244 @@
+    order = CONSTANT
+    family = MONOMIAL
+  []
+  # [disp_x]
+  # []
+  # [disp_y]
+  # []
```
Remove these comments.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,244 @@
+
+
+[Bounds]
+  # [irreversibility]
+  #   type = VariableOldValueBoundsAux
+  #   variable = bounds_dummy
+  #   bounded_variable = d
+  #   bound_type = lower
+  # []
```
Remove these comments.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,244 @@
+[]
+
+
+# [Postprocessors]
+#   # [extdriving]
+#   #   type = ElementAverageValue
+#   #   variable = 'ce'
+#   # []
+#   # # [beta_0]
+#   # #   type = ADElementAverageMaterialProperty
+#   # #   mat_prop = 'beta_0'
+#   # # []
+#   # [d_avg]
+#   #   type = AverageNodalVariableValue
+#   #   variable = d
+#   # []
+# []
```
Remove these comments.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,244 @@
+#   # []
+# []
+
+# [VectorPostprocessors]
+#   [damage]
+#     type = NodalValueSampler
+#     variable = 'd'
+#     sort_by = id
+#   []
+#   [ext]
+#     type = ElementValueSampler
+#     variable = 'ce'
+#     sort_by = id
+#   []
+# []
```
Remove these comments.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,244 @@
+[]
+
+[Outputs]
+  # [csv_]
+  #   type = CSV
+  #   file_base = surf_Gc2L0.35del1.16_frac
+  #   append_date = true
+  #   execute_vector_postprocessors_on = final
+  # []
```
Remove these comments.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+
+#pragma once
+
+#include "ADMaterial.h"
```
```suggestion
#include "Material.h"
```

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+
+#include "ADMaterial.h"
+
+class GeneralizedExternalDrivingForce : public ADMaterial
```
```suggestion
class GeneralizedExternalDrivingForce : public Material
```

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+protected:
+  virtual void computeQpProperties() override;
+
+  const MaterialPropertyName _ex_driving_name;
```
Add a doxygen string for each of the parameter here. For example, say
```
/// Name of the external driving energy
const MaterialPropertyName _ex_driving_name;
```

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+
+  const MaterialPropertyName _ex_driving_name;
+
+  ADMaterialProperty<Real> & _ex_driving;
```
Add a doxygen string.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  ADMaterialProperty<Real> & _ex_driving;
+
+  /// energy release rate
+  //const GenericMaterialProperty<Real,true> & _Gc;
+  const Real & _Gc;
+  // const Real & _Gc;
```
It should be a `const ADMaterialProperty<Real> &`

BoZeng1997 replied:
I am not computing based on qp. For now it can stay just a Real

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  // const Real & _Gc;
+
+  /// phase field regularization length
+  const Real & _L;
```
This should be a `const ADMaterialProperty<Real> &`.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  const Real & _L;
+  // const Real & _L;
+
+  const Real & _Lambda;
+
+  const Real & _mu;
```
Same here, they should be `const ADMaterialProperty<Real> &`

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  /// compressive strength
+  const Real & _sigma_cs;
+
+  ///
```
Missing comment.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  ///
+  const Real & _delta;
+
+  ///
```
Missing comments.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  Real _gamma_0;
+  Real _gamma_1;
+  Real _gamma_2;
+  // const ADVariableValue & _invar_1;
+  // const ADVariableValue & _invar_2;
+  // const ADMaterialProperty<Real> & _invar_1;
+  // const ADMaterialProperty<Real> & _invar_2;
```
Remove unused comments.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  // const ADVariableValue & _invar_2;
+  // const ADMaterialProperty<Real> & _invar_1;
+  // const ADMaterialProperty<Real> & _invar_2;
+  const MaterialPropertyName _rank_two_tensor;
```
What's this `_rank_two_tensor`?

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,62 @@
+  // const ADMaterialProperty<Real> & _invar_2;
+  const MaterialPropertyName _rank_two_tensor;
+  const ADMaterialProperty<RankTwoTensor> & _stress;
+// const GenericMaterialProperty<RankTwoTensor, is_ad> & _stress;
```
Remove unused comment.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+
+#include "GeneralizedExternalDrivingForce.h"
+#include "Function.h"
+// #include "RankTwoTensorImplementation.h"
+// #include "ADRankTwoTensorForward.h"
```
Remove unused comments.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+  InputParameters params = ADMaterial::validParams();
+  params.addClassDescription("computes the Kumar coeffs"
+                             "Ce()");
+//  params.addRequiredParam<Real>(
+//      "coef", "The coefficient describing the mismatch between the film and the substrate");
+//  params.addRequiredCoupledVar(
+//      "displacements",
+//      "The displacements appropriate for the simulation geometry and coordinate system");
```
Remove unused comments.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+GeneralizedExternalDrivingForce::validParams()
+{
+  InputParameters params = ADMaterial::validParams();
+  params.addClassDescription("computes the Kumar coeffs"
+                             "Ce()");
```
The line wrap here is weird, they should be in the same line.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+//  params.addRequiredCoupledVar(
+//      "displacements",
+//      "The displacements appropriate for the simulation geometry and coordinate system");
+  params.addRequiredParam<Real>(
+      "energy_release_rate", "energy release rate or fracture toughness");
```
```suggestion
  params.addParam<Real>(
      "energy_release_rate", "Gc", "energy release rate or fracture toughness");
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+//      "The displacements appropriate for the simulation geometry and coordinate system");
+  params.addRequiredParam<Real>(
+      "energy_release_rate", "energy release rate or fracture toughness");
+  params.addRequiredParam<Real>(
+      "phase_field_regularization_length", "phase_field_regularization length");
```
```suggestion
  params.addParam<Real>(
      "phase_field_regularization_length", "l", "The phase field regularization length");
```
Use a descriptive description here.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+      "energy_release_rate", "energy release rate or fracture toughness");
+  params.addRequiredParam<Real>(
+      "phase_field_regularization_length", "phase_field_regularization length");
+  params.addRequiredParam<Real>(
+      "Lame_first_parameter", "Lame's first parameter Lambda");
```
```suggestion
  params.addRequiredParam<Real>(
      "lambda", "Lame's first parameter Lambda");
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+      "phase_field_regularization_length", "phase_field_regularization length");
+  params.addRequiredParam<Real>(
+      "Lame_first_parameter", "Lame's first parameter Lambda");
+  params.addRequiredParam<Real>(
+      "shear_modulus", "shear modulus mu or G");
+  params.addRequiredParam<Real>(
+      "tensile_strength", "tensile strength");
+  params.addRequiredParam<Real>(
+      "compressive_strength", "compressive strength");
+  params.addRequiredParam<Real>(
+      "delta", "delta");
```
The line wraps are weird. Are you using clang-format?

BoZeng1997 replied:
this is the only way i can see all of them in half of the screen.

hugary1995 replied:
That's an issue with your editor. You should be able to configure the maximum column width within your editor. But the source code has to be clang-formated.

BoZeng1997 replied:
ok i will figure that out

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+      "rank_two_tensor","name of the stress tensor");
+  params.addParam<MaterialPropertyName>(
+      "external_driving_force_name", "ex_driving", "name of the material that holds the external_driving_force");
+//  params.addParam<MaterialPropertyName>(
+//      "first_invariant", "invar_1", "The first standard invariants of stress tensor");
+//  params.addParam<MaterialPropertyName>(
+//      "second_invariant", "invar_2", "The second standard invariants of stress tensor");
+// params.addCoupledVar("first_invariant","The first standard invariants of stress tensor");
+// params.addCoupledVar("second_invariant","The second standard invariants of stress tensor");
+  // params.addRequiredParam<ADMaterialProperty<Real>>("first_invariant","The first standard invariants of stress tensor");
+  // params.addRequiredParam<ADMaterialProperty<Real>>("second_invariant","The second standard invariants of stress tensor");
```
Remove these comments.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+      "compressive_strength", "compressive strength");
+  params.addRequiredParam<Real>(
+      "delta", "delta");
+  params.addRequiredParam<MaterialPropertyName>(
+      "rank_two_tensor","name of the stress tensor");
```
No need for this. I think the stress tensor is always called "stress".

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,128 @@
+
+GeneralizedExternalDrivingForce::GeneralizedExternalDrivingForce(const InputParameters & parameters)
+  : ADMaterial(parameters),
+    _ex_driving_name(getParam<MaterialPropertyName>("external_driving_force_name")),
+    _ex_driving(declareADProperty<Real>(_ex_driving_name)),
+    // _Gc(getADMaterialProperty<Real>("energy_release_rate")),
+    _Gc(getParam<Real>("energy_release_rate")),
+    _L(getParam<Real>("phase_field_regularization_length")),
+    _Lambda(getParam<Real>("Lame_first_parameter")),
+    _mu(getParam<Real>("shear_modulus")),
+    _sigma_ts(getParam<Real>("tensile_strength")),
+    _sigma_cs(getParam<Real>("compressive_strength")),
+    _delta(getParam<Real>("delta"))
+    ,_beta_0(declareADProperty<Real>("beta_0"))
+    ,_beta_1(declareADProperty<Real>("beta_1"))
+    ,_beta_2(declareADProperty<Real>("beta_2"))
+    ,_beta_3(declareADProperty<Real>("beta_3"))
+    // ,_invar_1(adCoupledValue("first_invariant"))
+    // ,_invar_2(adCoupledValue("second_invariant"))
+    // ,_invar_1(getADMaterialProperty<Real>("invar_1"))
+    // ,_invar_2(getADMaterialProperty<Real>("invar_2"))
+    ,_rank_two_tensor(getParam<MaterialPropertyName>("rank_two_tensor"))
+    ,_stress(getADMaterialProperty<RankTwoTensor>(_rank_two_tensor))
+    ,_F_surface(declareADProperty<Real>("F_surface"))
+    ,_J2(declareADProperty<Real>("J2"))
+//    _invar_1(getParam<Real>(getParam<MaterialPropertyName>("first_invariant"))),
+//    _invar_2(getParam<Real>(getParam<MaterialPropertyName>("second_invariant")))
+
+//    _E_int(declareADProperty<Real>(_E_int_name)),
+ //   _coef(getParam<Real>("coef")),
+  //  _ndisp(coupledComponents("displacements")),
+//    _disp(3)
```
Remove unused comments please.

hugary1995 reviewed (changes requested):
Overall looks good. I suggest we start from cleaning up some unused comments. Then we can double check the implementation.

BoZeng1997 commented:
I have cleaned unused comments and complete the necessary comments. But I did not make changes that will cause further changing in input deck. Were they changed, all of the input deck must be changed. I cannot guarantee a same result.

The problem right now is not with kumar's model, but j integral in general.

In the folder /benchmark/AT2surfing, there is input deck for surfing boundary problem under AT-2 without irreversibility (with irreversibility commented). The problem right now is that in the AT-2 model, with smaller mesh size, it still cannot return a close value to Gc.

BoZeng1997 commented:
... the number of changes made is still very large. I think it is because i rm your examples. I thought that is a way not to track the change in it.

hugary1995 commented:
The number of lines of addition is huge because you've checked in many exodus files -- remove those.
The number of lines of deletion is huge because you've removed everything in the benchmarks folder. We do plan to remove the entire benchmarks folder at some point in the future, but don't do that for now.

BoZeng1997 commented:
recovered benchmark examples.
cleared other redundancy.

hugary1995 commented:
Fantastic, thanks.

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,261 @@
+    phase_field = d
+    degradation_function = g
+    decomposition = NONE
+    output_properties = 'psie_active' #elastic_strain
```
`psie_active` is not the elastic strain.

BoZeng1997 replied:
I did not mean `psie_active` is `elastic_strain`. i commented `elastic_strain` to not output it.

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,261 @@
+    output_properties = 'stress'
+    outputs = exodus
+  []
+  [pk1]
+    type =RankTwoTensorMaterialConverter
+    ad_props_in = stress
+    ad_props_out = pk1_stress
+    intra_convert = true
+  []
```
Why do we need this converter?

BoZeng1997 replied:
because J integral object requires pk1_stress

hugary1995 replied:
The right thing to do here is to add an optional parameter in the `PhaseFieldJIntegral` postprocessor that asks for the name of the stress tensor, something like:
`params.addParam<MaterialPropertyName>("stress", "stress", "Name of the stress tensor");`

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,261 @@
+
+[Outputs]
+  exodus = true
+  file_base = surf_AT2_Gc2L0.35del1.16_dt1e-2_J5b_noirr_ela
```
I know you are probably doing a parameter study on your own branch. But let's choose a simpler file name for here.

BoZeng1997 replied:
got it

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,261 @@
+[Outputs]
+  exodus = true
+  file_base = surf_AT2_Gc2L0.35del1.16_dt1e-2_J5b_noirr_ela
+  append_date = true
```
Do not append date.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,193 @@
+  #   bounded_variable = d
+  #   bound_type = lower
+  # []
+  [lower]
+    type = ConstantBoundsAux
+    variable = bounds_dummy
+    bounded_variable = d
+    bound_type = lower
+    bound_value = 0
+  []
```
Why using a constant lower bound here?

BoZeng1997 replied:
In kumar's paper, they release the restriction on irreversibility until d = 0.95. I did not how to do that exactly in raccoon, so I first tried with no irreversibility at all since last Thursday.

hugary1995 replied:
I see. I would prefer to use the constraint in its "correct" form. Then if we see issues with that, we fix it (instead of using someone's "engineering" approach).

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,193 @@
+    order = CONSTANT
+    family = MONOMIAL
+  []
+  [ce]
+    order = CONSTANT
+    family = MONOMIAL
+  []
```
ce isn't used anywhere.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,193 @@
+[]
+
+[Outputs]
+  exodus = true
+  file_base = surf_AT2_Gc2L0.35del1.16_dt1e-2_J5b_noirr_frac
```
Do not generate output from the subapp.

hugary1995 commented on benchmarks/AT2surfing/fracture.i:
```diff
@@ -0,0 +1,193 @@
+[Outputs]
+  exodus = true
+  file_base = surf_AT2_Gc2L0.35del1.16_dt1e-2_J5b_noirr_frac
+  append_date = true
```
Do not append date, as that will make it impossible to do regression tests.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,58 @@
+  ADMaterialProperty<Real> & _ex_driving;
+
+  /// energy release rate
+  const Real & _Gc;
```
I know you were probably using a `const Real &` for your own testing purposes. But it should be a `const ADMaterialProperty<Real> &` in the master branch to be consistent with all other objects.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,58 @@
+  /// phase field regularization length
+  const Real & _L;
+  /// Lame's first parameter
+  const Real & _Lambda;
+  /// shear modulus
+  const Real & _mu;
```
Same thing here. They should be `const ADMaterialProperty<Real> &`

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,58 @@
+  /// regularization length dependent parameter
+  const Real & _delta;
+
+  /// values computed during
+  ADMaterialProperty<Real> & _beta_0;
+  ADMaterialProperty<Real> & _beta_1;
+  ADMaterialProperty<Real> & _beta_2;
+  ADMaterialProperty<Real> & _beta_3;
+  Real _gamma_0;
+  Real _gamma_1;
+  Real _gamma_2;
```
values computed during ______?

hugary1995 replied:
I don't think they need to be defined as members. They are quantities you compute at every qp, and you don't need to store them as material properties, because they won't be used by anything else, and they won't be output for visualization.

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,58 @@
+  Real _gamma_2;
+  /// name of the degraded stress tensor
+  const MaterialPropertyName _rank_two_tensor;
+  /// tje degraded stress tensor
```
```suggestion
  /// The stress
```

hugary1995 commented on include/materials/GeneralizedExternalDrivingForce.h:
```diff
@@ -0,0 +1,58 @@
+  /// tje degraded stress tensor
+  const ADMaterialProperty<RankTwoTensor> & _stress;
+  /// strength surface computed
+  ADMaterialProperty<Real> & _F_surface;
+  /// j2 invariant
+  ADMaterialProperty<Real> & _J2;
```
Same comment here: I don't think they should be declared as members and stored as material properties.

hugary1995 commented on include/postprocessors/PhaseFieldJIntegral.h:
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "SideIntegralPostprocessor.h"
+#include "RankTwoTensor.h"
```
```suggestion
#include "RankTwoTensorForward.h"
```

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,261 @@
+  [Jint]
+    type = PhaseFieldJIntegral
+    J_direction = '1 0 0'
+    elastic_energy_name = psie_active
```
Most likely, the error in your J-R curve comes from here. Here you need to pass in the strain energy density, i.e. `g * psie_active + psie_inactive`, instead of just the active strain energy density.

BoZeng1997 replied:
So in my case, since i am not using decomposition, it should be `psie = g * psie_active`

hugary1995 replied:
Two material properties are output from the constitutive model: `psie` and `psie_active`, where `psie = g * psie_active + psie_inactive`. So in your case, you can just use `psie`.

BoZeng1997 replied:
with psie, J2 = 9.6e-2, 5% error from Gc = 9.1e-2. the error is smaller, but is it small enough?

hugary1995 replied:
I think it's fairly small, but it doesn't really make sense to compare just one number, we need to look at the JR curve as a function of crack extension.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,85 @@
+    _sigma_ts(getParam<Real>("tensile_strength")),
+    _sigma_cs(getParam<Real>("compressive_strength")),
+    _delta(getParam<Real>("delta"))
+    ,_beta_0(declareADProperty<Real>("beta_0"))
+    ,_beta_1(declareADProperty<Real>("beta_1"))
+    ,_beta_2(declareADProperty<Real>("beta_2"))
+    ,_beta_3(declareADProperty<Real>("beta_3"))
+    ,_rank_two_tensor(getParam<MaterialPropertyName>("rank_two_tensor"))
+    ,_stress(getADMaterialProperty<RankTwoTensor>(_rank_two_tensor))
+    ,_F_surface(declareADProperty<Real>("F_surface"))
+    ,_J2(declareADProperty<Real>("J2"))
```
Please enforce clang-format, commas cannot be the first character of a line. Let me know if you need help setting up clang-format.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,85 @@
+    ,_F_surface(declareADProperty<Real>("F_surface"))
+    ,_J2(declareADProperty<Real>("J2"))
+{
+  _gamma_0 = _sigma_ts/6.0/(3.0*_Lambda+2.0*_mu)+_sigma_ts/6.0/_mu;
+  _gamma_1 = (1.0+_delta)*(_sigma_cs-_sigma_ts)/(2.0*_sigma_ts*_sigma_cs);
+  _gamma_2 = (1.0+_delta)*(_sigma_cs+_sigma_ts)/(2.0*_sigma_ts*_sigma_cs)*std::sqrt(3.0);
```
They should be computed at each qp, because `lambda` and `mu` can be inhomogeneous.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,85 @@
+  ADReal I1 = _stress[_qp](0,0)+_stress[_qp](1,1)+_stress[_qp](2,2);
+  ADReal I2 = _stress[_qp](0,0)*_stress[_qp](1,1)+_stress[_qp](0,0)*_stress[_qp](2,2)+_stress[_qp](1,1)*_stress[_qp](2,2)-pow(_stress[_qp](0,1),2)-pow(_stress[_qp](1,2),2)-pow(_stress[_qp](2,0),2);
+  _J2[_qp] = ( pow(_stress[_qp](0,0)-_stress[_qp](1,1),2)+pow(_stress[_qp](1,1)-_stress[_qp](2,2),2)+pow(_stress[_qp](0,0)-_stress[_qp](2,2),2) )/6.0+pow(_stress[_qp](0,1),2)+pow(_stress[_qp](0,2),2)+pow(_stress[_qp](2,1),2);
+  try{
+    if (_J2[_qp]<0)
+    {throw 11;}
+  }
+  catch(int e){
+    std::cout<<"Negative J2 "<<_J2[_qp]<<std::endl;
+    exit(e);
+  }
```
```suggestion
  mooseAssert(_J2[_qp] >= 0, "Some messages here if the assertion fails");
```
But I don't think you need this assertion anyways, it's always gonna be nonnegative.

Also, this is not the usual way of computing J2. J2 is simply the norm of the deviatoric part of the stress.

BoZeng1997 replied:
How to compute J2? like:
RankTwoTensor stress_deviatoric = stress.deviatoric();
J2 = stress_deviatoric.doubleContraction(stress_deviatoric)/2;

BoZeng1997 replied:
Will this assert catches an imaginary part?

hugary1995 replied:
You can use
```
ADRankTwoTensor stress_dev = stress.deviatoric();
ADReal J2 = stress_dev.doubleContraction(stress_dev);
```
You can derive it yourself, i.e. expand the double contraction and see if it is the same expression as your original code.

hugary1995 replied:
> Will this assert catches an imaginary part?

`Real` doesn't contain an imaginary part. Also a norm can never be imaginary.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,85 @@
+    std::cout<<"Negative J2 "<<_J2[_qp]<<std::endl;
+    exit(e);
+  }
+  _beta_0[_qp]= _delta*_temp;
+  _beta_1[_qp] = -_gamma_1*_temp+_gamma_0;
+  _beta_2[_qp] = -_gamma_2*_temp+std::sqrt(3.0)*_gamma_0;
+  _beta_3[_qp] = (1-std::sqrt(I1*I1)/I1)*(_J2[_qp]/2.0/_mu+I1*I1/6.0/(3.0*_Lambda+2.0*_mu));
```
Again, these `beta`s don't need to be material properties. Just declare them as `Real` within this scope.

BoZeng1997 replied:
They were the product of once me trying to fix the negative J2 problem. i will remove them.

hugary1995 reviewed (changes requested):
We are getting close. I think I found the cause of your J-R curve issue.

Also, please enforce clang-format -- most long equations are pretty unreadable right now. If you need help setting up clang-format just let me know.

hugary1995 commented:
Here is a decision for you to make: if you only want to demonstrate the usage of J-integral in this PR, then all you need are the source code for the J-integral postprocessor and one tutorial input file. If you also want to check in the Kumar model, then in addition to the source code that you have checked in, I also need to see at least regression test/tutorial the demonstrate its usage.

I'm fine with both, it's up to you.

hugary1995 commented:
You can see why the tests are failing if you click on `details`. For example, the `Precheck` is currently because:
```
##########################################################################
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
Merge branch 'JintSurfing' of https://github.com/BoZeng1997/raccoon into test
clear
recover examples
Jintsurfing-clean2
AT2surfing-clean
for AT2 Jintegral Surfing only
GeneralizedKumar_issue_Jint
all kinds of kumar for test
lastindirectkumar
modifyGeneralizedExternalDrivingForce to have ce in elasticity.i
recover tutorials from upstream
recover tutorials
20210623_updateKumar_ProblemNAN
first_extdriving

ERROR: The following files contain banned keywords (std::cout, std::cerr, sleep, print_trace):
	src/materials/GeneralizedExternalDrivingForce.C
##########################################################################
```

BoZeng1997 commented:
what issue should it be linked to? Tutorial?

BoZeng1997 commented:
what is this ticket reference exactly? is not mine #88 ?

BoZeng1997 commented:
COPY FROM MOOSE WEBPAGE:

Referencing Issues
Every modification to MOOSE must reference an issue number. This means that every Pull Request (PR) that flows into MOOSE must have contain at least one commit that references an issue relevant to what you are working on (e.g. refs #<number> (where <number> is an issue number on the MOOSE GitHub issue page, such as #1234). If your PR completely addresses an issue, you can automatically close it by prepending "closes" or "fixes" to the issue reference (e.g., closes #1234). Issue numbers are automatically checked by our testing system.

hugary1995 commented:
If you don't find a relevant issue, you can create your own. Then you can modify your commit message to reference that issue.

BoZeng1997 commented:
> Here is a decision for you to make: if you only want to demonstrate the usage of J-integral in this PR, then all you need are the source code for the J-integral postprocessor and one tutorial input file. If you also want to check in the Kumar model, then in addition to the source code that you have checked in, I also need to see at least one regression test/tutorial that demonstrates its usage.
> 
> I'm fine with both, it's up to you.

For example, is the input deck already committed in this folder enough to be a tutorial input file for simply J integral? 

I do also want to check the kumar model, since I have already committed the object and it cannot return a Jintegral value close to that of AT2. I want to know what kind of test is good enough as a regression test. can surfing boundary value problem serve as a regression test? we know the expected speed of the crack propagation but the result need to be obtained from paraview.

hugary1995 commented:
> For example, is the input deck already committed in this folder enough to be a tutorial input file for simply J integral?

Yes.

> I do also want to check the kumar model, since I have already committed the object and it cannot return a Jintegral value close to that of AT2. I want to know what kind of test is good enough as a regression test. can surfing boundary value problem serve as a regression test? we know the expected speed of the crack propagation but the result need to be obtained from paraview.

You might have misunderstood the meaning of regression test. Regression test is used to ensure consistency, not for correctness.

Take a look at other files in the tutorials folder, each tutorial has its own `tests` specification. Most of them are simply diffing the output against some `gold` files you check in. This is to make sure future changes to raccoon do not break you stuff.

BoZeng1997 commented:
> > For example, is the input deck already committed in this folder enough to be a tutorial input file for simply J integral?
> 
> Yes.
> 
> > I do also want to check the kumar model, since I have already committed the object and it cannot return a Jintegral value close to that of AT2. I want to know what kind of test is good enough as a regression test. can surfing boundary value problem serve as a regression test? we know the expected speed of the crack propagation but the result need to be obtained from paraview.
> 
> You might have misunderstood the meaning of regression test. Regression test is used to ensure consistency, not for correctness.
> 
> Take a look at other files in the tutorials folder, each tutorial has its own `tests` specification. Most of them are simply diffing the output against some `gold` files you check in. This is to make sure future changes to raccoon do not break you stuff.

I did misunderstand the regression test.

But for now, since we are not updating our moose, is one tutorial file, say kumar's model with surfing BVP,  enough for kumar's usage?

hugary1995 commented:
Yes, that's sufficient. It will demonstrate the usage of both the J integral post processor and the nucleation model.

hugary1995 commented:
Just ping me when you think this PR is ready for review.

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,266 @@
+  []
+[]
+
+[BCs]
+  [bottom_y]
+    type = FunctionDirichletBC
+    variable = disp_y
+    boundary = bottom_half_bottom
+    function = bc_func
+  []
+  [top_y]
+    type = FunctionDirichletBC
+    variable = disp_y
+    boundary = top_half_top
+    function = bc_func
+  []
+[]
```
This is not quite right. The surfing BC should be applied to all of the external boundaries (excluding crack surfaces), and should be applied to both `disp_x` and `disp_y`.

BoZeng1997 replied:
Do you mean it should be also applied to the (outer) left, top, right, and bottom, and on two directions for each boundary?

BoZeng1997 replied:
In kumar's paper, they were only applying disp_y on top and bottom.

hugary1995 replied:
That's something to discuss tomorrow for sure.

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,266 @@
+[]
+
+[Postprocessors]
+  [Fy]
+    type = NodalSum
+    variable = fy
+    boundary = top_half_top
+  []
```
Reaction force isn't needed in this example.

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,266 @@
+    displacements = 'disp_x disp_y'
+    boundary = 'top_half_left top_half_top top_half_right bottom_half_right bottom_half_bottom bottom_half_left'
+  []
+[]
```
Add a postprocessor to integrate the crack surface density.

BoZeng1997 replied:
it is in material

BoZeng1997 replied:
`[Materials]`

hugary1995 replied:
Yes, I can see that. I am saying you should integrate the crack surface density "gamma" which is computed in the material into a scalar representing the total crack surface area. Something like
```
[a]
  type = ADElementIntegralMaterialProperty
  mat_prop = gamma
[]
```

hugary1995 replied:
I guess you were integrating that in paraview, which isn't unreliable.

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,266 @@
+    type = ADStressDivergenceTensors
+    variable = disp_y
+    component = 1
+    save_in = fy
```
reaction force isn't needed in this example.

hugary1995 commented on benchmarks/AT2surfing/elasticity.i:
```diff
@@ -0,0 +1,266 @@
+[]
+
+[AuxVariables]
+  [fy]
+  []
```
Reaction force isn't needed in this example.

hugary1995 reviewed (changes requested):
Caught a few more minor things.

BoZeng1997 commented:
It is ready for review now. I have made modifications to most of the issues mentioned above. I still keep the J2 and F_surface in the object GeneralizedExternalForce because it will be useful for debuging. They will not be stored in the future. The surfing boundary condition did not change.

BoZeng1997 commented:
I am not sure it will pass the checks. I need to run for the bus now

hugary1995 commented:
Thanks, this is getting really close now. The "Travis CI" test fails because of documentation failure -- some classes are missing documentations.

Let's discuss about the boundary condition in tomorrow's group meeting.

A few minor things:
- Now that you've added that `benchmarks/surfingBVP` example, the `benchmarks/AT2surfing` example becomes redundant, so lets remove that one.
- We shouldn't be adding any new examples into that `benchmarks` folder -- those are our old examples. We want to migrate everything to the new `tutorials` folder. So could you move your `benchmakrs/surfingBVP` example into the tutorials folder? I think that will be a very useful tutorial for everyone. Also, let's try to keep it consistent with the naming convention in that folder, so instead of `surfingBVP`, let's rename it to something like `surfing_BVP` with an underscore.
- One last thing you need to do is to tell the testing system how this new tutorial should be tested. To do that, you need to add a file called `tests` in the folder of your example. In that `tests` file, you should describe how the test should run. You can simply copy from another `tests` from an existing tutorial. Make sure you only run it for one or two steps so that the test can finish within 20 seconds.
- One other thing you could do, although not necessary, is to document the tutorial. To document a tutorial, first add an entry to the menu in `doc/menu.yml`, something like `12. Steady crack propagation: tutorials/12_surfing.md`. Then add the actual documentation page at `doc/content/tutorials/12_surfing.md`. You can copy from existing documentations, e.g. `01_small_deformation_elasticity.md`, as a starting point.

As a reminder, to build documentation on your own machine:
- commit everything, make sure everything is tracked by git
- compile your application, make sure it is up-to-date
- `cd doc`
- `./moosedocs.py build --serve`, this will build the documentation and serve the website at your localhost.

hugary1995 commented:
Hey Bo, have you had a chance to address those comments? If you need help with anything just let me know.

BoZeng1997 commented:
I was not on it this week. My Mac machine became slow again and I have to focus on one task. It is slow even just typing on Github page. My laptop is back and I am preparing it back to work. I will be back on this soon.

hugary1995 commented:
Okay no worries -- wasn't pushing you.

BoZeng1997 commented:
Do you accept a kumar's  that might be wrongly implemented? I am still looking for the reason I cannot have Jintegral = Gc.

hugary1995 commented:
It's not related to the nucleation model, is it? You've shown that J isn't exactly equal to Gc even with standard AT1 or AT2.

hugary1995 commented:
But I do remember that when I tried the surfing problem, I got J = Gc for a purely elastic case. Let me play with it and see if I can reproduce what I had before.

BoZeng1997 commented:
I think it may have to do with the difference in the 2020 formulation and the 2021 formulation. What I have pushed in is the 2021 version. The 2020 version gave a result Jint/Gc = 0.95 which is good. But the 2021 version has improved behavior on compression( said in paper). I need to decide which pair of formulation and example to push in. The ones you saw in github is the 2021 formulation with 2020 example (they probably do not match).

BoZeng1997 commented:
not for review yet

BoZeng1997 commented:
> I see why the `Test` test is failing for you. That's because you committed a new submodule moose. Please don't do that.

How do I remove this redundant moose safely?

BoZeng1997 commented:
> Yes, you should worry about those errors. These errors tell you exactly what files are missing. These errors will prevent the Documentation test from passing, and a PR cannot be merged with a failing test.
> 
> > The ./moosedocs.py build --serve process almost stuck at [I 210827 14:41:40 handlers:64] Start detecting changes . Is this normal?
> 
> It is not stuck, it is detecting changes. You should be able to view the documentation website locally at your localhost.

How long should this step usually take?

BoZeng1997 commented:
```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   doc/menu.yml
        modified:   moose (new commits)
        modified:   tutorials/surfing_boundary_problem/test

no changes added to commit (use "git add" and/or "git commit -a")
```
I cannot restore the moose (new commit)

BoZeng1997 commented:
I install the most up-to-date clang format 11 on my Mac machine and tried clang-format the src code of GeneralizedExternalDrivingForce.C. The result is the same as the one which got rejected by this clang-format check in the first place. So it was not the problem of my clang-format version. There has to be some setting in clang format that should not be default value. I also tried format by hand and it solved most of the problem. 

It is ready for review.

hugary1995 commented on include/auxkernels/ConditionalBoundsAux.h:
```diff
@@ -0,0 +1,30 @@
+
+#include "BoundsAuxBase.h"
+
+// Forward Declarations
+class ConditionalBoundsAux;
```
I don't think we need a forward declaration here.

hugary1995 commented on include/auxkernels/ConditionalBoundsAux.h:
```diff
@@ -0,0 +1,30 @@
+// Forward Declarations
+class ConditionalBoundsAux;
+
+template <>
+InputParameters validParams<ConditionalBoundsAux>();
```
This is the deprecated way of declaring `validParams`. Just remove it.

hugary1995 commented on src/auxkernels/ConditionalBoundsAux.C:
```diff
@@ -0,0 +1,41 @@
+
+registerMooseObject("raccoonApp", ConditionalBoundsAux);
+
+defineLegacyParams(ConditionalBoundsAux);
```
Remove this line.

hugary1995 commented on src/auxkernels/ConditionalBoundsAux.C:
```diff
@@ -0,0 +1,41 @@
+{
+  InputParameters params = BoundsAuxBase::validParams();
+  params.addClassDescription(
+      "Provides the bound of the variable a lower bound according to variable "
+      "value"
+      "for kumar's phase-field fracture model 2020"
+      "variable to PETSc's SNES variational inequalities solver.");
```
```suggestion
      "This class conditionally enforces a bound. When the variable value is below a given threshold, a constant value is used as the bound; when the variable value is above a given threshold, irreversibility is enforced.");
```

hugary1995 commented on src/auxkernels/ConditionalBoundsAux.C:
```diff
@@ -0,0 +1,41 @@
+  params.addRequiredParam<Real>("fixed_bound_value", "The value of fixed bound for the variable");
+  params.addRequiredParam<Real>("threshold_value",
+                                "The threshold for conditional history bound for the variable");
+  return params;
```
You are also making a strong assumption in this class that this is imposing a lower bound. This is fine, but you need to suppress the corresponding parameter, for example
```
params.set<MooseEnum>("bound_type") = "lower";
params.suppressParameter<bool>("bound_type");
```

hugary1995 commented on src/auxkernels/ConditionalBoundsAux.C:
```diff
@@ -0,0 +1,41 @@
+{
+  Real d_old = _var.getNodalValueOld(*_current_node);
+  if (d_old >= _threshold_value)
+  {
+    return d_old;
+  }
+  else
+  {
+    return _fixed_bound_value;
+  }
```
When there is only a single statement within the curly braces, you should remove the curly braces.
```suggestion
    return d_old;
  else
    return _fixed_bound_value;
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+
+InputParameters GeneralizedExternalDrivingForce::validParams() {
+  InputParameters params = Material::validParams();
+  params.addClassDescription("computes the Kumar coeffs"
+                             "Ce()");
```
```suggestion
  params.addClassDescription("computes the external driving force for nucleation given a strength envelope. The implementation follows Kumar et. al. (2020).");
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  params.addClassDescription("computes the Kumar coeffs"
+                             "Ce()");
+  params.addParam<MaterialPropertyName>(
+      "energy_release_rate", "Gc", "energy release rate or fracture toughness");
```
I think we use `fracture_toughness` everywhere in RACCOON. So let's be consistent.
```suggestion
      "fracture_toughness", "Gc", "energy release rate or fracture toughness");
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  params.addParam<MaterialPropertyName>(
+      "energy_release_rate", "Gc", "energy release rate or fracture toughness");
+  params.addParam<MaterialPropertyName>(
+      "phase_field_regularization_length", "l",
```
I think we use `regularization_length` everywhere in RACCOON.
```suggestion
      "regularization_length", "l",
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  params.addParam<MaterialPropertyName>(
+      "phase_field_regularization_length", "l",
+      "the phase field regularization length");
+  params.addParam<MaterialPropertyName>("Lambda", "Lambda",
```
Should be lower case `lambda`.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  params.addParam<MaterialPropertyName>("shear_modulus", "G",
+                                        "shear modulus mu or G");
+  params.addRequiredParam<Real>("tensile_strength",
+                                "critical tensile strength");
```
```suggestion
                                "The tensile strength of the material beyond which the material fails.");
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  params.addRequiredParam<Real>("tensile_strength",
+                                "critical tensile strength");
+  params.addRequiredParam<Real>("compressive_strength",
+                                "critical compressive strength");
```
```suggestion
                                "The compressive strength of the material beyond which the material fails.");
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+                                "critical tensile strength");
+  params.addRequiredParam<Real>("compressive_strength",
+                                "critical compressive strength");
+  params.addRequiredParam<Real>("delta", "delta");
```
Please add an actually descriptive description for `delta`.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  params.addRequiredParam<Real>("compressive_strength",
+                                "critical compressive strength");
+  params.addRequiredParam<Real>("delta", "delta");
+  params.addParam<MaterialPropertyName>("rank_two_tensor", "stress",
```
Don't call it `rank_two_tensor`. I think it is better to call this parameter `stress`:
```suggestion
  params.addParam<MaterialPropertyName>("stress", "stress",
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  params.addRequiredParam<Real>("delta", "delta");
+  params.addParam<MaterialPropertyName>("rank_two_tensor", "stress",
+                                        "name of the stress tensor");
+  params.addParam<MaterialPropertyName>(
+      "external_driving_force_name", "ex_driving",
+      "name of the material that holds the external_driving_force");
```
Let's try to be consistent with parameter naming. For example, we use the parameter `fracture_toughness` instead of `fracture_toughness_name` for the name of the fracture toughness. Similarly, I think we should name this parameter `external_driving_force` instead of `external_driving_force_name`.

BoZeng1997 replied:
No, this is just the name of that parameter, not the parameter itself.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  ADReal _gamma_1 = (1.0 + _delta) / (2.0 * _sigma_ts * _sigma_cs);
+  ADReal _gamma_2 =
+      (8 * _mu[_qp] + 24 * K - 27 * _sigma_ts) / 144 / _mu[_qp] / K;
+  ADReal _temp = _Gc[_qp] * 3.0 / _L[_qp] / 8.0;
```
No, I won't let you hard code this. The normalization constant `c0` isn't always `8/3`.

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+  ADReal _gamma_2 =
+      (8 * _mu[_qp] + 24 * K - 27 * _sigma_ts) / 144 / _mu[_qp] / K;
+  ADReal _temp = _Gc[_qp] * 3.0 / _L[_qp] / 8.0;
+  ADReal I1 = _stress[_qp](0, 0) + _stress[_qp](1, 1) + _stress[_qp](2, 2);
```
```suggestion
  ADReal I1 = _stress[_qp].trace();
```

hugary1995 commented on src/materials/GeneralizedExternalDrivingForce.C:
```diff
@@ -0,0 +1,88 @@
+      (8 * _mu[_qp] + 24 * K - 27 * _sigma_ts) / 144 / _mu[_qp] / K;
+  ADReal _temp = _Gc[_qp] * 3.0 / _L[_qp] / 8.0;
+  ADReal I1 = _stress[_qp](0, 0) + _stress[_qp](1, 1) + _stress[_qp](2, 2);
+  ADReal J2 = (pow(_stress[_qp](0, 0) - _stress[_qp](1, 1), 2) +
+               pow(_stress[_qp](1, 1) - _stress[_qp](2, 2), 2) +
+               pow(_stress[_qp](0, 0) - _stress[_qp](2, 2), 2)) /
+                  6.0 +
+              pow(_stress[_qp](0, 1), 2) + pow(_stress[_qp](0, 2), 2) +
+              pow(_stress[_qp](2, 1), 2);
```
```suggestion
  ADRankTwoTensor stress_dev = _stress[_qp].deviatoric();
  ADReal J2 = stress_dev.doubleContraction(stress_dev);
```

hugary1995 reviewed (changes requested):
Some minor naming questions and code suggestions.

hugary1995 commented:
I just pushed three commits cleaning up everything. Take a look at the changes I made -- if you are fine with those then this PR is good to go.

I was mostly making minor style changes. But one mistake I noticed is that you named the test specification `test` instead of `tests`, and as a result the regression test wasn't being run. I've fixed that for you.

hugary1995 commented:
I would also like to thank you for your contribution! It is hard for everyone to make their first pull request. Hopefully you've learnt something from this PR. You should definitely be proud of this work!

hugary1995 commented:
@milljm This is the PR waiting for CIVET tests.

BoZeng1997 commented:
Yeah, I think those modifications are great. Thank you for being **so patient** with this PR! I did learned a lot!
